app/main/helpers/frameworks.py

```diff
@@ -311,3 +311,26 @@ def check_agreement_is_related_to_supplier_framework_or_abort(agreement, supplie
         abort(404)
     if not agreement.get('frameworkSlug') or agreement.get('frameworkSlug') != supplier_framework.get('frameworkSlug'):
         abort(404)
+
+
+def get_frameworks_closed_and_open_for_applications(frameworks):
+
+    all_open_fwks = []
+    all_opening_fwks = []
+    all_closed_fwks = []
+    framework_frameworks = set([framework['framework'] for framework in frameworks])
+
+    # We need to find one framework iteration per framework-framework, open > coming > closed
+    for f_f in framework_frameworks:
+        framework_iterations = [framework for framework in frameworks if framework['framework'] == f_f]
+        open_fwks = [fwk for fwk in framework_iterations if fwk['status'] == 'open']
+        opening_fwks = [fwk for fwk in framework_iterations if fwk['status'] == 'coming']
+        closed_fwks = [fwk for fwk in framework_iterations if fwk['status'] not in ('open', 'coming')]
+        if open_fwks:
+            all_open_fwks.append(open_fwks[0])
+        elif opening_fwks:
+            all_opening_fwks.append(opening_fwks[0])
+        else:
+            all_closed_fwks.append(closed_fwks[0])
+
+    return all_open_fwks, all_opening_fwks, all_closed_fwks
```

Comment:
It should probably be commented that the output of this implicitly depends on the ordering of `frameworks`.

Reply:
Clearer *to me* would be e.g.

```
all_open_fwks = tuple(chain.from_iterable(
    islice(grp, 1)  # only want the first of each metaframework
    for _, grp in groupby(
        sorted(
            (fw for fw in frameworks if fw["status"] in ("open",)),
            key=lambda fw: fw["framework"],
        ),
        key=lambda fw: fw["framework"],
    )
))
```

the `("open",)` could then be replaced by function argument and this could generate all three sequences.

Reply:
(or maybe a lambda to be able to cope with the `not in`)

Reply:
I find this much more confusing, but would be good to factor out to a function, I'll see what I can do.

Reply:
Actually I that because we only one one framework per metaframework across all three lists this needs to be an atomic operation that generates all three lists in parallel.
If we have a function like this for each status (or set of statuses) then we then additionally need a whole bunch of extra logic to whittle it down (if there's a g-cloud framework in list one then remove all g-clouds from lists 2 and 3, if there's a dos framework in list 2 then remove any dos frameworks from list 3, etc.)

Reply:
Well, it's plugging together a few tools to get a result - you can tell approximately what it's trying to do by its use of `groupby` and `sorted`, where the alternative is just "well, it throws around some lists and stuff, picks some values out and then something comes out the end". To me it's like the difference between reading assembly and reading a high-level expression.

Reply:
ohhhhright.

Reply:
still, easy to misread this style of code.

Reply:
I'm trying to figure out a neat way to do this still... don't hold your breath.

Reply:
Well, it's quite a simple change to my method:

```
def status_priority(status):
    if status == "open":
        return 0
    elif status not in ...:
        return 1
    else:
        return 2

displayed_fwks = tuple(chain.from_iterable(
    islice(grp, 1)  # only want the first of each metaframework
    for _, grp in groupby(
        sorted(
            (fw for fw in frameworks if fw["status"] in ("open",)),
            key=lambda fw: (fw["framework"], status_priority(fw["status"])),
        ),
        key=lambda fw: fw["framework"],
    )
))
```

Then you don't even have to `chain` the results back together in the view. Because that's what this function is doing - it's a prioritization, a sort, with some rules over uniqueness which are easily doable using `groupby`

Reply:
But if you have a single list of "displayed_frameworks" the logic in the view needs to be completely different, and deal with six possible framework states.

Reply:
Oh, they're displayed in yet another different order, aren't they. Well, that's just another sorting, and it's quite easy to filter out just the statuses you want.

Anyway, this feels like a discussion for mon.

Reply:
```all_open_fwks = tuple(chain.from_iterable(
    islice(grp, 1)  # only want the first of each metaframework
    for _, grp in groupby(
        sorted(
            (fw for fw in frameworks if fw["status"] in ("open",)),
            key=lambda fw: fw["framework"],
        ),
        key=lambda fw: fw["framework"],
    )
))```

x_x

Reply:
I've pushed an updated version of this, Finds the frameworks your way and then splits the list out to pass the separate things into the template, which keeps the template logic the same as it was.  Seems OK to me but see what you reckon?